src/others/pells_equation.md

```diff
@@ -0,0 +1,84 @@
+Therefore, $ ( x_{0} + \sqrt{d} \cdot y_{0} )^{n}  < u + v \cdot \sqrt{d} < ( x_{0} + \sqrt{d} \cdot y_{0} )^{n+1}$
+
+Multiplying the above inequality by $( x_{0} - \sqrt{d} \cdot y_{0} )^{n}$,(which is > 0 and < 1) we get
+
+$ 1 < (u + v \cdot \sqrt{d})( x_{0} - \sqrt{d} \cdot y_{0} )^{n} < ( x_{0} + \sqrt{d} \cdot y_{0} ) $
+Because both $(u + v \cdot \sqrt{d})$ and $( x_{0} - \sqrt{d} \cdot y_{0} )^{n}$ have norm 1, their product is also a solution.
+But this contradicts our assumption that $( x_{0} + \sqrt{d} \cdot y_{0} )$ is the smallest solution. Therefore, there is no solution between $( x_{0} + \sqrt{d} \cdot y_{0} )^{n}$ and $( x_{0} + \sqrt{d} \cdot y_{0} )^{n+1}$.
+
+Hence, we conclude that all solutions are given by $( x_{0} + \sqrt{d} \cdot y_{0} )^{n}$ for some integer $n$.
+
+## To find the smallest positive solution
+### Expressing the solution in terms of continued fractions
+We can express the solution in terms of continued fractions. The continued fraction of $\sqrt{d}$ is periodic. Let's assume the continued fraction of $\sqrt{d}$ is $[a_{0}; \overline{a_{1}, a_{2}, \ldots, a_{r}}]$. The smallest positive solution is given by the convergent $[a_{0}; a_{1}, a_{2}, \ldots, a_{r}]$ where $r$ is the period of the continued fraction.
+
+The convergents $p_{n}/q_{n}$ are the rational approximations to $\sqrt{d}$ obtained by truncating the continued fraction expansion at each stage. These convergents can be computed recursively.
+
+Check whether the convergent satisfies the Pell's equation. If it does, then the convergent is the smallest positive solution.
+
+Let's take an example to understand this by solving the equation $x^{2} - 2 \cdot y^{2} = 1$.
+$\sqrt{2} = [1; \overline{2}] = 1 + 1/(2 + 1/(2 + 1/(2+ ...))) $. The convergents are $1/1, 3/2, 7/5, 17/12, 41/29, 99/70, \ldots$.
+Now check for each convergent whether it satisfies the Pell's equation. The smallest positive solution is $3/2$.
+
+### How to calculate the continued fraction of $\sqrt{d}$?
+Let's find the continued fraction of $\def\sf{\sqrt 7}\sf$.
+
+$\sf \approx 2.6457 = 2 + 0.6457$
+
+$a_{0} = 2 $
+
+Subtract $a_{0}$ from the number and take the reciprocal of the remaining number.
+
+That is, we calculate ${1\over \sf - 2} \approx 1.5486$. The integer part $a_{1}$ is $1$.
+So:
+    $$\sf=2+\cfrac{1}{1+\cfrac1{\vdots}}$$ Where we haven't calculated the $\vdots$ part yet.
+To get that, we subtract $a_{1}$ from the number and take the reciprocal of the remaining number. That is, we calculate ${1\over 1.5486 - 1} \approx 1.8228$. The integer part $a_{2}$ is $1$.
+So:
+    $$\sf=2+\cfrac{1}{1+\cfrac1{1+\cfrac1{\vdots}}}$$
+Now ${1\over 1.8228 - 1} \approx 1.2153$. So $a_{3} = 1$.
+Continuing this process, ${1\over 1.2153 - 1} \approx 4.645$. So $a_{4} = 4$.
+
+$$\sf=2+\cfrac{1}{1+\cfrac1{1+\cfrac1{1+\cfrac4{\vdots}}}}$$
+
+we get the continued fraction of $\sf$ as $[2; 1, 1, 4, 1, 1, 4, \ldots]$.
```

Comment:
Note this relies on floating point numbers. You can do it entirely with integer arithmetic.

Reply:
Can you give me reference for doing it using integer arithmetic completely?

Reply:
"Quadratic irrationality" section of [continued fraction article](https://cp-algorithms.com/algebra/continued-fractions.html) has it:

```py
# compute the continued fraction of sqrt(n)
def sqrt(n):
    n0 = math.floor(math.sqrt(n))
    x, y, z = 1, 0, 1
    a = []
    def step(x, y, z):
        a.append((x * n0 + y) // z)
        t = y - a[-1]*z
        x, y, z = -z*x, z*t, t**2 - n*x**2
        g = math.gcd(x, math.gcd(y, z))
        return x // g, y // g, z // g

    used = dict()
    for i in range(n):
        used[x, y, z] = i
        x, y, z = step(x, y, z)
        if (x, y, z) in used:
            return a
```

Reply:
included as reference